### ISSUE TITLE
Read files from memory

### ISSUE THREAD

**Issue by frodofine** [chemfiles/chemfiles#219]
For speed and other uses (see #188), I believe it would be useful to read files from a location in memory instead of only from disk.  A hackish version of this is implemented in #187, but there are many ways to improve support.  No matter what implementation is chosen, new constructors are going to be implemented for each format.

The best method would use a C++ concept like streams and pass this object to formats to perform operations on instead of a file path.  The format factory class would then be responsible for opening the stream, moving the file operations out of the formats themselves.  A potential issue here is how to handle formats like `NetCDF` or those handled by `Molfiles` (maybe addressed with #154?). A *messy* solution could be to have two maps of formats, one for those which support this 'stream' method and a second one for legacy formats that require file paths.

**Comment by Luthaf** [comment]
I have been thinking about this lately, so I wanted to do a bit of a brain dump =)

This is something that would be very nice to have, we mostly need to figure out how to do it given our set of constrains. I can see the following constrains:

1. Reading from a memory buffer is easy enough, since the buffer already exists and is full. Writing inside an user-provided memory buffer could be much harder: what happens when the buffer is full? Either we need a way to request more memory from the caller (passing callbacks for example), or we want to manager the buffer ourself, and then have a way for the caller to retrieve it.
2. The interface should work not only with native C++ objects, but also within the C API, and map nicely (as much as possible) to the other languages bindings. This unfortunately means that C++-style stream objects might be hard to use.

> A potential issue here is how to handle formats like NetCDF or those handled by Molfiles (maybe addressed with #154?).

I think we are fine by not supporting in-memory reading and writing for binary formats, if the underlying library do not support it. At the same time, we want to be able to provide in-memory read and write capacities if the library supports it.

----

On the API side, I think that in-memory I/O is different enough from file I/O to have its own interface, at least for `Trajectory`. So either a new subclass/sibling class to `Trajectory` (`MemoryTrajectory`/`BufferTrajectory`/`StringTrajectory`), or few new functions added to trajectory:

```cpp
// C++ API

class MemoryBuffer: TextFile {
    // This class would both implement the TextFile interface and be usable
    // by binary formats as a std::vector<char> like storage
}

class Trajectory {
    /// new constructor using a MemoryBuffer, format can be used to 
    /// specify format and compression. 
    /// mode='a' is not supported
    /// mode='r' takes a full buffer and read from it
    /// mode='w' takes an empty buffer and write to it
    Trajectory(MemoryBuffer buffer, char mode, std::string format);
 
    /// instead of a constructor, this can also be static function/a free functions
    /// removing the need to expose MemoryBuffer publicly
    static Trajectory memory_reader(const char* data, size_t size, std::string format);
    static Trajectory memory_writer(std::string format);

    /// this would only work for in-memory trajectories, to access the buffer after 
    /// writing
    optional<string_view> buffer() const;
}

// C API
CHFL_TRAJECTORY* chfl_trajectory_mem_reader(const char* data, size_t size, const char* format);

CHFL_TRAJECTORY* chfl_trajectory_mem_writer(const char* format);

chfl_status chfl_trajectory_buffer(const CHFL_TRAJECTORY* traj, const char** data, size_t* size);
```


Any though, uses-cases this might be missing ?

**Comment by frodofine** [comment]
I'm working on this now, I'll update the PR sometime this week.

**Comment by frodofine** [comment]
As far as the API you suggested is concerned, I implemented it almost exactly as you specified. I think the static functions are a better choice as exposing `MemoryBuffer` publically will lead to the API being more complex for the end-user as they would need to initialize the buffer first. This way, they only need to pass a `char*` and size in the case of reading, or nothing in the case of writing.  Adding additional constructors is also not a good idea as `Trajector(char*, File::Mode, File::Compression)` would be ambiguous with the current constructor. I did use a private constructor to implement the static functions, however. From what I can tell, the `MemoryBuffer.hpp` file will need to be whitelisted as it is needed for the `unique_ptr` deconstructor. An alternative would be to use a raw pointer and `Trajectory`'s deconstructor, but I think whitelisting the file (and potentially moving it to a different directory) is what is best.

As for the backend, I subclassed `TextFileImpl` instead of `TextFile` and used the C++ std::stream library to implement the required functionality. Note that I used two classes, one for reading memory and another for writing memory. I think this makes more sense than a single class for both as it segregates the functionality in a nice way. The only thing that I think should be changed is how formats that do not handle reading/writing to memory report the error (the `has_memory()` is not elegant). Also, the names of some functions are not very good (eg `gzfile_in_place`).

When all is done, I think I will squash the commits. The history is quite long as it is.
